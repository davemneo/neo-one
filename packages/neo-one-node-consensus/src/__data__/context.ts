import { common } from '@neo-one/client-common';
import _ from 'lodash';
import { cloneBlockSent, cloneInitial, Context } from '../context';
import { keys } from './keys';

const validators = keys.map(({ publicKey }) => publicKey);
const previousHash = common.stringToUInt256('0xd42561e3d30e15be6400b6df2f328e02d2bf6354c41dce433bc57687c82144bf');

const expectedView = _.range(0, 7).map(() => 0);
const blockReceivedTimeSeconds = 1513018010;

const viewNumber = 0;
const primaryIndex = 1;
const blockIndex = primaryIndex;
const backupIndex = 2;
const backupContext = new Context({
  type: 'backup',
  previousHash,
  blockIndex,
  viewNumber,
  myIndex: backupIndex,
  primaryIndex,
  expectedView,
  validators,
  blockReceivedTimeSeconds,
});

const backupPrivateKey = keys[backupIndex].privateKey;

const primaryContext = new Context({
  type: 'primary',
  previousHash,
  blockIndex,
  viewNumber,
  myIndex: primaryIndex,
  primaryIndex,
  expectedView,
  validators,
  blockReceivedTimeSeconds,
});

const primaryPrivateKey = keys[primaryIndex].privateKey;

const blockSentBackupContext = cloneBlockSent(backupContext);
const initialBackupContext = cloneInitial(backupContext, {
  type: 'backup',
  viewNumber,
  primaryIndex,
});

const blockSentPrimaryContext = cloneBlockSent(primaryContext);
const initialPrimaryContext = cloneInitial(primaryContext, {
  type: 'primary',
  viewNumber,
  primaryIndex,
});

export const context = {
  backupPrivateKey,
  primaryPrivateKey,
  previousHash,
  blockSentBackupContext,
  initialBackupContext,
  blockSentPrimaryContext,
  initialPrimaryContext,
};
