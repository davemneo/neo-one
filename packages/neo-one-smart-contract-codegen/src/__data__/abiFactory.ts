import { ABI, ABIDefault, ABIReturn } from '@neo-one/client-common';

// tslint:disable-next-line:no-any
const getCombinations = (list: ReadonlyArray<ReadonlyArray<any>>) => {
  const floorsIn = list.reduceRight<ReadonlyArray<number>>((acc, right) => [right.length * acc[0]].concat(acc), [1]);
  const max = floorsIn[0];
  const floors = floorsIn.slice(1);

  const iteratorFuncs = list.map((item, index) =>
    index === list.length - 1
      ? (value: number) => item[value % item.length]
      : (value: number) => item[Math.floor(value / floors[index]) % item.length],
  );

  return [...Array(max)].map((_, index) => iteratorFuncs.map((func) => func(index)));
};

const abiFunctionOptions: ReadonlyArray<{ readonly [key in string]?: boolean }> = [
  { constant: true },
  { send: true },
  { sendUnsafe: true },
  { sendUnsafe: true, receive: true },
  { receive: true },
  { claim: true },
  { refundAssets: true },
  { completeSend: true },
  {},
];

const abiParameterOptions: ReadonlyArray<{ readonly [key in string]?: boolean | ABIDefault }> = [
  { default: { type: 'sender' } },
  { rest: true },
  { default: { type: 'sender' }, rest: true },
  {},
];

const abiReturns: ReadonlyArray<ABIReturn> = [
  {
    type: 'ForwardValue',
  },
  {
    type: 'Boolean',
    optional: true,
  },
  {
    type: 'Boolean',
    forwardedValue: true,
  },
  {
    type: 'Integer',
    decimals: 4,
  },
  {
    type: 'Array',
    value: {
      type: 'Boolean',
    },
  },
  {
    type: 'Map',
    key: { type: 'Integer', decimals: 0 },
    value: { type: 'Boolean' },
  },
  {
    type: 'Object',
    properties: {
      property: { type: 'Boolean' },
    },
  },
];

const boolReturnType: ABIReturn = { type: 'Boolean' };

const createBaseABIFunctions = (): ReadonlyArray<ABI> => {
  const combinations = getCombinations([abiReturns, abiFunctionOptions]);

  return combinations.map(([abiReturn, abiFuncProps]) => {
    const funcPropNames = Object.keys(abiFuncProps);
    const name = `${abiReturn.type}${`${funcPropNames.length === 0 ? '' : '_'}${funcPropNames.join('_')}`}`;

    return {
      functions: [
        {
          name,
          returnType: abiReturn,
          ...abiFuncProps,
        },
      ],
    };
  });
};

const createBaseABIEvents = (): ReadonlyArray<ABI> => {
  const combinations = getCombinations([abiReturns, abiParameterOptions]);

  return combinations.map(([abiReturn, abiParamProps]) => {
    const paramPropNames = Object.keys(abiParamProps);
    const name = `${abiReturn.type}${`${paramPropNames.length === 0 ? '' : '_'}${paramPropNames.join('_')}`}`;

    return {
      functions: [
        {
          name: 'foo',
          returnType: boolReturnType,
        },
      ],
      events: [
        {
          name,
          parameters: [
            {
              name: 'arg',
              ...abiReturn,
              ...abiParamProps,
            },
          ],
        },
      ],
    };
  });
};

const createBaseABIFunctionParameters = (): ReadonlyArray<ABI> => {
  const combinations = getCombinations([abiReturns, abiParameterOptions]);

  return combinations.map(([abiReturn, abiParamProps]) => {
    const paramPropNames = Object.keys(abiParamProps);
    const name = `${abiReturn.type}${`${paramPropNames.length === 0 ? '' : '_'}${paramPropNames.join('_')}`}`;

    return {
      functions: [
        {
          name,
          parameters: [
            {
              name: 'arg',
              ...abiReturn,
              ...abiParamProps,
            },
          ],
          returnType: boolReturnType,
        },
      ],
    };
  });
};

export const abiFactory = {
  createBaseABIFunctions,
  createBaseABIFunctionParameters,
  createBaseABIEvents,
};
