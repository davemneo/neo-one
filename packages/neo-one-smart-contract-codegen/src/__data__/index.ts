export * from './nep5';
export * from './testUtils';
export * from './abiFactory';
