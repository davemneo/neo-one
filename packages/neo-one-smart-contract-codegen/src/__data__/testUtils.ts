import { ABI } from '@neo-one/client-common';
import { genSmartContractTypes } from '../types';

const testABI = (abi: ABI, name: string) => {
  test(name, () => {
    expect({
      inputABI: abi,
      types: genSmartContractTypes(name, abi),
    }).toMatchSnapshot();
  });
};

export const testUtils = {
  testABI,
};
