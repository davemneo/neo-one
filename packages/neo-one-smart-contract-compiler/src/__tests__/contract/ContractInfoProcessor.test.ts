import { helpers } from '../../__data__';
import { DiagnosticCode } from '../../DiagnosticCode';

const properties = `
public readonly properties = {
  codeVersion: '1.0',
  author: 'dicarlo2',
  email: 'alex.dicarlo@neotracker.io',
  description: 'The TestSmartContract',
};
`;

describe('ContractInfoProcessor', () => {
  test('allowed types', async () => {
    const node = await helpers.startNode();
    await node.addContract(`
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
      }
    `);
  });

  test('no symbol properties', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        public readonly [Symbol.iterator]: string = 'foo';
      }
    `,
      { type: 'error' },
    );
  });

  test('no computed properties', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        public readonly ['foo']: string = 'foo';
      }
    `,
      { type: 'error' },
    );
  });

  test('no computed methods', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        public ['foo'](): string {
          return 'foo';
        }
      }
    `,
      { type: 'error' },
    );
  });

  test('no static properties', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        public static readonly foo = 'foo';
      }
    `,
      { type: 'error' },
    );
  });

  test('no static methods', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        public static foo() {
          return 'foo';
        }
      }
    `,
      { type: 'error' },
    );
  });

  test('no method decorators', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      function dec(target: any, propertyKey: string, descriptor: PropertyDescriptor): void {
        // do nothing
      }

      export class TestSmartContract extends SmartContract {
        ${properties}

        @dec
        public foo() {
          return 'foo';
        }
      }
    `,
      { type: 'error' },
    );
  });

  test('no method parameter decorators', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      function dec(target: Object, propertyKey: string | symbol, parameterIndex: number): void {
        // do nothing
      }

      export class TestSmartContract extends SmartContract {
        ${properties}

        public foo(@dec value: string) {
          return 'foo';
        }
      }
    `,
      { type: 'error' },
    );
  });

  test('no get accessor method decorators', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      function dec(target: any, propertyKey: string, descriptor: PropertyDescriptor): void {
        // do nothing
      }

      export class TestSmartContract extends SmartContract {
        ${properties}
        private x: string = 'foo';

        @dec
        public get foo(): string {
          return this.x;
        }

        public set foo(value: string) {
          this.x = value;
        }
      }
    `,
      { type: 'error' },
    );
  });

  test('no set accessor method decorators', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      function dec(target: any, propertyKey: string, descriptor: PropertyDescriptor): void {
        // do nothing
      }

      export class TestSmartContract extends SmartContract {
        ${properties}
        private x: string = 'foo';

        public get foo(): string {
          return this.x;
        }

        @dec
        public set foo(value: string) {
          this.x = value;
        }
      }
    `,
      { type: 'error' },
    );
  });

  test('no set accessor parameter decorators', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      function dec(target: Object, propertyKey: string | symbol, parameterIndex: number): void {
        // do nothing
      }

      export class TestSmartContract extends SmartContract {
        ${properties}
        private x: string = 'foo';

        public get foo(): string {
          return this.x;
        }

        public set foo(@dec value: string) {
          this.x = value;
        }
      }
    `,
      { type: 'error' },
    );
  });

  test('no deploy method', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        public deploy(): void {
          // do nothing
        }
      }
    `,
      { type: 'error' },
    );
  });

  test('multiple smart contracts', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
      }

      export class TestSmartContract2 extends SmartContract {
        ${properties}
      }
    `,
      { type: 'error' },
    );
  });

  test('abstract smart contracts', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export abstract class TestSmartContract extends SmartContract {
        ${properties}
      }
    `,
      { type: 'error' },
    );
  });

  test('new SmartContract()', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
      }

      new TestSmartContract();
    `,
      { type: 'error' },
    );
  });

  test('@receive with incorrect return type', () => {
    helpers.compileString(
      `
      import { SmartContract, receive } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        @receive
        public verify(): number {
          return 20;
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractMethod },
    );
  });

  test('@receive with @constant', () => {
    helpers.compileString(
      `
      import { SmartContract, receive, constant } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        @receive
        @constant
        public verify(): boolean {
          return true;
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractMethod },
    );
  });

  test('@sendUnsafe with @constant', () => {
    helpers.compileString(
      `
      import { SmartContract, sendUnsafe, constant } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        @sendUnsafe
        @constant
        public verify(): boolean {
          return true;
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractMethod },
    );
  });

  test('@claim with @constant', () => {
    helpers.compileString(
      `
      import { SmartContract, claim, constant } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        @claim
        @constant
        public verify(): boolean {
          return true;
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractMethod },
    );
  });

  test('structured storage set in constructor', () => {
    helpers.compileString(
      `
      import { SmartContract, MapStorage } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        private readonly foo: MapStorage<string, string>;
        public constructor() {
          super();
          this.foo = MapStorage.for<string, string>();
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidStructuredStorageFor },
    );
  });

  test('structured storage set public', () => {
    helpers.compileString(
      `
      import { SmartContract, MapStorage } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        public readonly foo = MapStorage.for<string, string>();
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidStructuredStorageFor },
    );
  });

  test('structured storage private modifiable', () => {
    helpers.compileString(
      `
      import { SmartContract, MapStorage } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        private foo = MapStorage.for<string, string>();
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidStructuredStorageFor },
    );
  });

  test('structured storage protected abstract', () => {
    helpers.compileString(
      `
      import { SmartContract, MapStorage } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        protected abstract readonly foo: MapStorage<string, string>;
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidStructuredStorageFor },
    );
  });

  test('invalid storage type', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      class Foo {}

      export class TestSmartContract extends SmartContract {
        ${properties}
        private readonly foo: Foo = new Foo();
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractStorageType },
    );
  });

  test('invalid storage structured array type', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      class Foo {}

      export class TestSmartContract extends SmartContract {
        ${properties}
        private readonly foo: Array<Foo> = [new Foo()];
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractStorageType },
    );
  });

  test('invalid storage structured map type', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      class Foo {}

      export class TestSmartContract extends SmartContract {
        ${properties}
        private readonly foo: Array<Map<string, Foo>> = [new Map<string, Foo>().set('foo', new Foo())];
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractStorageType },
    );
  });

  test('invalid storage structured set type', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      class Foo {}

      export class TestSmartContract extends SmartContract {
        ${properties}
        private readonly foo: Set<Map<string, Foo>> = new Set([new Map<string, Foo>().set('foo', new Foo())]);
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractStorageType },
    );
  });

  test('invalid property function not readonly', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        private foo = () => {
          // do nothing
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractStorageType },
    );
  });

  test('invalid property function set in constructor', () => {
    helpers.compileString(
      `
      import { SmartContract } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}
        private readonly foo: () => void;

        public constructor() {
          super();
          this.foo = () => {
            // do nothing
          }
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractStorageType },
    );
  });

  test('invalid @sendUnsafe method with ForwardValue', () => {
    helpers.compileString(
      `
      import { SmartContract, ForwardValue, sendUnsafe } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        @sendUnsafe
        public foo(value: ForwardValue): boolean {
          return true;
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractType },
    );
  });

  test('invalid @receive method with ForwardValue', () => {
    helpers.compileString(
      `
      import { SmartContract, ForwardValue, receive } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        @receive
        public foo(value: ForwardValue): boolean {
          return true;
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractType },
    );
  });

  test('invalid @claim method with ForwardValue', () => {
    helpers.compileString(
      `
      import { SmartContract, ForwardValue, claim } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        @claim
        public foo(value: ForwardValue): boolean {
          return true;
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractType },
    );
  });

  test('invalid @sendUnsafe method with ForwardedValue', () => {
    helpers.compileString(
      `
      import { SmartContract, ForwardedValue, sendUnsafe } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        @sendUnsafe
        public foo(value: ForwardedValue<number>): boolean {
          return true;
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractType },
    );
  });

  test('invalid @receive method with ForwardedValue', () => {
    helpers.compileString(
      `
      import { SmartContract, ForwardedValue, receive } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        @receive
        public foo(value: ForwardedValue<number>): boolean {
          return true;
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractType },
    );
  });

  test('invalid @claim method with ForwardedValue', () => {
    helpers.compileString(
      `
      import { SmartContract, ForwardedValue, claim } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        @claim
        public foo(value: ForwardedValue<number>): boolean {
          return true;
        }
      }
    `,
      { type: 'error', code: DiagnosticCode.InvalidContractType },
    );
  });
});
