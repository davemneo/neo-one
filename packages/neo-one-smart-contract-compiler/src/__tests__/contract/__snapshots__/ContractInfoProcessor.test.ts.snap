// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractInfoProcessor @claim with @constant 1`] = `
"snippetCode.ts (15,9): SmartContract methods marked with @send, @sendUnsafe, @receive or @claim cannot be @constant.

      13 | 
      14 |         @claim
    > 15 |         @constant
         |         ^
      16 |         public verify(): boolean {
      17 |           return true;
      18 |         }
"
`;

exports[`ContractInfoProcessor @receive with @constant 1`] = `
"snippetCode.ts (15,9): SmartContract methods marked with @send, @sendUnsafe, @receive or @claim cannot be @constant.

      13 | 
      14 |         @receive
    > 15 |         @constant
         |         ^
      16 |         public verify(): boolean {
      17 |           return true;
      18 |         }
"
`;

exports[`ContractInfoProcessor @receive with incorrect return type 1`] = `
"snippetCode.ts (14,9): SmartContract methods marked with @send, @sendUnsafe, @receive or @claim must return a boolean.

      12 | 
      13 | 
    > 14 |         @receive
         |         ^
      15 |         public verify(): number {
      16 |           return 20;
      17 |         }
"
`;

exports[`ContractInfoProcessor @sendUnsafe with @constant 1`] = `
"snippetCode.ts (15,9): SmartContract methods marked with @send, @sendUnsafe, @receive or @claim cannot be @constant.

      13 | 
      14 |         @sendUnsafe
    > 15 |         @constant
         |         ^
      16 |         public verify(): boolean {
      17 |           return true;
      18 |         }
"
`;

exports[`ContractInfoProcessor abstract smart contracts 1`] = `
"snippetCode.ts (4,7): Cannot compile abstract SmartContract.

      2 |       import { SmartContract } from '@neo-one/smart-contract';
      3 | 
    > 4 |       export abstract class TestSmartContract extends SmartContract {
        |       ^
      5 |         
      6 | public readonly properties = {
      7 |   codeVersion: '1.0',
"
`;

exports[`ContractInfoProcessor invalid @claim method with ForwardValue 1`] = `
"snippetCode.ts (15,20): Cannot use ForwardValue or ForwardedValue as parameters for methods marked with @send, @sendUnsafe, @receive or @claim.

      13 | 
      14 |         @claim
    > 15 |         public foo(value: ForwardValue): boolean {
         |                    ^
      16 |           return true;
      17 |         }
      18 |       }
"
`;

exports[`ContractInfoProcessor invalid @claim method with ForwardedValue 1`] = `
"snippetCode.ts (15,20): Cannot use ForwardValue or ForwardedValue as parameters for methods marked with @send, @sendUnsafe, @receive or @claim.

      13 | 
      14 |         @claim
    > 15 |         public foo(value: ForwardedValue<number>): boolean {
         |                    ^
      16 |           return true;
      17 |         }
      18 |       }
"
`;

exports[`ContractInfoProcessor invalid @receive method with ForwardValue 1`] = `
"snippetCode.ts (15,20): Cannot use ForwardValue or ForwardedValue as parameters for methods marked with @send, @sendUnsafe, @receive or @claim.

      13 | 
      14 |         @receive
    > 15 |         public foo(value: ForwardValue): boolean {
         |                    ^
      16 |           return true;
      17 |         }
      18 |       }
"
`;

exports[`ContractInfoProcessor invalid @receive method with ForwardedValue 1`] = `
"snippetCode.ts (15,20): Cannot use ForwardValue or ForwardedValue as parameters for methods marked with @send, @sendUnsafe, @receive or @claim.

      13 | 
      14 |         @receive
    > 15 |         public foo(value: ForwardedValue<number>): boolean {
         |                    ^
      16 |           return true;
      17 |         }
      18 |       }
"
`;

exports[`ContractInfoProcessor invalid @sendUnsafe method with ForwardValue 1`] = `
"snippetCode.ts (15,20): Cannot use ForwardValue or ForwardedValue as parameters for methods marked with @send, @sendUnsafe, @receive or @claim.

      13 | 
      14 |         @sendUnsafe
    > 15 |         public foo(value: ForwardValue): boolean {
         |                    ^
      16 |           return true;
      17 |         }
      18 |       }
"
`;

exports[`ContractInfoProcessor invalid @sendUnsafe method with ForwardedValue 1`] = `
"snippetCode.ts (15,20): Cannot use ForwardValue or ForwardedValue as parameters for methods marked with @send, @sendUnsafe, @receive or @claim.

      13 | 
      14 |         @sendUnsafe
    > 15 |         public foo(value: ForwardedValue<number>): boolean {
         |                    ^
      16 |           return true;
      17 |         }
      18 |       }
"
`;

exports[`ContractInfoProcessor invalid property function not readonly 1`] = `
"snippetCode.ts (13,9): Invalid SmartContract storage type. See https://neo-one.io/docs/properties-and-storage#Storage-Types for valid storage types.

      11 | };
      12 | 
    > 13 |         private foo = () => {
         |         ^
      14 |           // do nothing
      15 |         }
      16 |       }
"
`;

exports[`ContractInfoProcessor invalid property function set in constructor 1`] = `
"snippetCode.ts (13,9): Invalid SmartContract storage type. See https://neo-one.io/docs/properties-and-storage#Storage-Types for valid storage types.

      11 | };
      12 | 
    > 13 |         private readonly foo: () => void;
         |         ^
      14 | 
      15 |         public constructor() {
      16 |           super();
"
`;

exports[`ContractInfoProcessor invalid storage structured array type 1`] = `
"snippetCode.ts (15,9): Invalid SmartContract storage type. See https://neo-one.io/docs/properties-and-storage#Storage-Types for valid storage types.

      13 | };
      14 | 
    > 15 |         private readonly foo: Array<Foo> = [new Foo()];
         |         ^
      16 |       }
      17 |     
"
`;

exports[`ContractInfoProcessor invalid storage structured map type 1`] = `
"snippetCode.ts (15,9): Invalid SmartContract storage type. See https://neo-one.io/docs/properties-and-storage#Storage-Types for valid storage types.

      13 | };
      14 | 
    > 15 |         private readonly foo: Array<Map<string, Foo>> = [new Map<string, Foo>().set('foo', new Foo())];
         |         ^
      16 |       }
      17 |     
"
`;

exports[`ContractInfoProcessor invalid storage structured set type 1`] = `
"snippetCode.ts (15,9): Invalid SmartContract storage type. See https://neo-one.io/docs/properties-and-storage#Storage-Types for valid storage types.

      13 | };
      14 | 
    > 15 |         private readonly foo: Set<Map<string, Foo>> = new Set([new Map<string, Foo>().set('foo', new Foo())]);
         |         ^
      16 |       }
      17 |     
"
`;

exports[`ContractInfoProcessor invalid storage type 1`] = `
"snippetCode.ts (15,9): Invalid SmartContract storage type. See https://neo-one.io/docs/properties-and-storage#Storage-Types for valid storage types.

      13 | };
      14 | 
    > 15 |         private readonly foo: Foo = new Foo();
         |         ^
      16 |       }
      17 |     
"
`;

exports[`ContractInfoProcessor multiple smart contracts 1`] = `
"snippetCode.ts (15,7): Multiple exported SmartContracts in one file is not supported.

      13 |       }
      14 | 
    > 15 |       export class TestSmartContract2 extends SmartContract {
         |       ^
      16 |         
      17 | public readonly properties = {
      18 |   codeVersion: '1.0',
"
`;

exports[`ContractInfoProcessor new SmartContract() 1`] = `
"snippetCode.ts (15,7): Cannot construct a SmartContract

      13 |       }
      14 | 
    > 15 |       new TestSmartContract();
         |       ^
      16 |     
"
`;

exports[`ContractInfoProcessor no computed methods 1`] = `
"snippetCode.ts (13,16): SmartContracts cannot have computed methods or properties.

      11 | };
      12 | 
    > 13 |         public ['foo'](): string {
         |                ^
      14 |           return 'foo';
      15 |         }
      16 |       }
"
`;

exports[`ContractInfoProcessor no computed properties 1`] = `
"snippetCode.ts (13,25): SmartContracts cannot have computed methods or properties.

      11 | };
      12 | 
    > 13 |         public readonly ['foo']: string = 'foo';
         |                         ^
      14 |       }
      15 |     
"
`;

exports[`ContractInfoProcessor no deploy method 1`] = `
"snippetCode.ts (14,16): deploy is a reserved method name in SmartContracts.

      12 | 
      13 | 
    > 14 |         public deploy(): void {
         |                ^
      15 |           // do nothing
      16 |         }
      17 |       }
"
`;

exports[`ContractInfoProcessor no get accessor method decorators 1`] = `
"snippetCode.ts (19,9): Custom decorators are not supported

      17 |         private x: string = 'foo';
      18 | 
    > 19 |         @dec
         |         ^
      20 |         public get foo(): string {
      21 |           return this.x;
      22 |         }
"
`;

exports[`ContractInfoProcessor no method decorators 1`] = `
"snippetCode.ts (18,9): Custom decorators are not supported

      16 | 
      17 | 
    > 18 |         @dec
         |         ^
      19 |         public foo() {
      20 |           return 'foo';
      21 |         }
"
`;

exports[`ContractInfoProcessor no method parameter decorators 1`] = `
"snippetCode.ts (18,20): Custom decorators are not supported

      16 | 
      17 | 
    > 18 |         public foo(@dec value: string) {
         |                    ^
      19 |           return 'foo';
      20 |         }
      21 |       }
"
`;

exports[`ContractInfoProcessor no set accessor method decorators 1`] = `
"snippetCode.ts (23,9): Custom decorators are not supported

      21 |         }
      22 | 
    > 23 |         @dec
         |         ^
      24 |         public set foo(value: string) {
      25 |           this.x = value;
      26 |         }
"
`;

exports[`ContractInfoProcessor no set accessor parameter decorators 1`] = `
"snippetCode.ts (23,24): Custom decorators are not supported

      21 |         }
      22 | 
    > 23 |         public set foo(@dec value: string) {
         |                        ^
      24 |           this.x = value;
      25 |         }
      26 |       }
"
`;

exports[`ContractInfoProcessor no static methods 1`] = `
"snippetCode.ts (13,16): SmartContracts cannot have static methods or properties.

      11 | };
      12 | 
    > 13 |         public static foo() {
         |                ^
      14 |           return 'foo';
      15 |         }
      16 |       }
"
`;

exports[`ContractInfoProcessor no static properties 1`] = `
"snippetCode.ts (13,16): SmartContracts cannot have static methods or properties.

      11 | };
      12 | 
    > 13 |         public static readonly foo = 'foo';
         |                ^
      14 |       }
      15 |     
"
`;

exports[`ContractInfoProcessor no symbol properties 1`] = `
"snippetCode.ts (13,25): SmartContracts cannot have computed methods or properties.

      11 | };
      12 | 
    > 13 |         public readonly [Symbol.iterator]: string = 'foo';
         |                         ^
      14 |       }
      15 |     
"
`;

exports[`ContractInfoProcessor structured storage private modifiable 1`] = `
"snippetCode.ts (13,9): Persistent storage classes may only be used as SmartContract private or protected readonly properties with a property initializer.

      11 | };
      12 | 
    > 13 |         private foo = MapStorage.for<string, string>();
         |         ^
      14 |       }
      15 |     
"
`;

exports[`ContractInfoProcessor structured storage protected abstract 1`] = `
"snippetCode.ts (13,9): Persistent storage classes may only be used as SmartContract private or protected readonly properties with a property initializer.

      11 | };
      12 | 
    > 13 |         protected abstract readonly foo: MapStorage<string, string>;
         |         ^
      14 |       }
      15 |     
"
`;

exports[`ContractInfoProcessor structured storage set in constructor 1`] = `
"snippetCode.ts (13,9): Persistent storage classes may only be used as SmartContract private or protected readonly properties with a property initializer.

      11 | };
      12 | 
    > 13 |         private readonly foo: MapStorage<string, string>;
         |         ^
      14 |         public constructor() {
      15 |           super();
      16 |           this.foo = MapStorage.for<string, string>();
"
`;

exports[`ContractInfoProcessor structured storage set public 1`] = `
"snippetCode.ts (13,9): Persistent storage classes may only be used as SmartContract private or protected readonly properties with a property initializer.

      11 | };
      12 | 
    > 13 |         public readonly foo = MapStorage.for<string, string>();
         |         ^
      14 |       }
      15 |     
"
`;
